pallets/subtensor/src/subnets/leasing.rs

```diff
@@ -0,0 +1,382 @@
+        let end_block = lease.end_block.ok_or(Error::<T>::LeaseHasNoEndBlock)?;
+        ensure!(now >= end_block, Error::<T>::LeaseHasNotEnded);
+
+        // Transfer ownership to the beneficiary
+        ensure!(
+            Self::coldkey_owns_hotkey(&lease.beneficiary, &hotkey),
+            Error::<T>::BeneficiaryDoesNotOwnHotkey
+        );
+        SubnetOwner::<T>::insert(lease.netuid, lease.beneficiary.clone());
+        Self::set_subnet_owner_hotkey(lease.netuid, &hotkey);
+
+        // Stop tracking the lease coldkey and hotkey
+        let _ = frame_system::Pallet::<T>::dec_providers(&lease.coldkey).defensive();
+        let _ = frame_system::Pallet::<T>::dec_providers(&lease.hotkey).defensive();
+
+        // Remove the lease, its contributors and accumulated dividends from storage
+        let clear_result =
+            SubnetLeaseShares::<T>::clear_prefix(lease_id, T::MaxContributors::get(), None);
+        AccumulatedLeaseDividends::<T>::remove(lease_id);
+        SubnetLeases::<T>::remove(lease_id);
+
+        // Remove the beneficiary proxy
+        T::ProxyInterface::remove_lease_beneficiary_proxy(&lease.coldkey, &lease.beneficiary)?;
+
+        Self::deposit_event(Event::SubnetLeaseTerminated {
+            beneficiary: lease.beneficiary,
+            netuid: lease.netuid,
+        });
+
+        if clear_result.unique < T::MaxContributors::get() {
+            // We have cleared less than the max number of shareholders, so we need to refund the difference
+            let k = clear_result.unique.into();
+            Ok(Some(
+                Weight::from_parts(56_635_122, 6148)
+                    .saturating_add(Weight::from_parts(912_993, 0).saturating_mul(k))
+                    .saturating_add(T::DbWeight::get().reads(4_u64))
+                    .saturating_add(T::DbWeight::get().reads((1_u64).saturating_mul(k)))
+                    .saturating_add(T::DbWeight::get().writes(6_u64))
+                    .saturating_add(T::DbWeight::get().writes((1_u64).saturating_mul(k)))
+                    .saturating_add(Weight::from_parts(0, 2529).saturating_mul(k)),
+            )
+            .into())
+        } else {
+            // We have cleared the max number of shareholders, so we don't need to refund anything
+            Ok(().into())
+        }
+    }
+
+    /// Hook used when the subnet owner's cut is distributed to split the amount into dividends
+    /// for the contributors and the beneficiary in shares relative to their initial contributions.
+    ///
+    /// It will ensure the subnet has enough alpha in its liquidity pool before swapping it to tao to be distributed,
+    /// and if not enough liquidity is available, it will accumulate the dividends for later distribution.
+    pub fn distribute_leased_network_dividends(lease_id: LeaseId, owner_cut_alpha: u64) {
+        // Ensure the lease exists
+        let lease = match SubnetLeases::<T>::get(lease_id) {
```

Comment:
It seems "let-else" is appropriate here: https://doc.rust-lang.org/rust-by-example/flow_control/let_else.html